### PR TITLE
📖 Add ADR around use of structured proxy types

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
kind/documentation

**What this PR does / why we need it**:

ADR around the benefits of using structured proxy types

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Adresses #113

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits into logical changes
- [x] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: docs/adr/0007-use-structured-proxy-types.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
...rnal/rancher` directory, to generate deep copy definitions and therefore allow specified resources to match the `Object` [interface](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/object.go#L45) provided by the controller-runtime. This will allow us to operate independently from `Rancher` specified types, without sacrificing functionality provided by the `Object` interface, including but not limited to a variety of helper methods. This also allows us to avoid unnecessary overhead in converting our objects to unstructured, inconsistent behaviors happening around setting API types on these objects and allows us to use the `envtest` and avoid using the `fakeClient` while testing (see the [issue](https://github.com/kubernetes-sigs/controller-runtime/issues/2308)) and use regular [setup procedures](https://kubebuilder.io/reference/envtest.html?highlight=envtest#writing-tests) for envtest around type registration. This will also allow us to simplify the process of writing e2e [tests](../../test/e2e/README.md#testing) and would allow us to know in advance if the specification of proxy type diverges from `Rancher` spec, using the latest `Rancher` version in e2e.
```
Comment (furkatgofurov7) [inline docs/adr/0007-use-structured-proxy-types.md:24]: Can we make a list of "allows us features" with bullet points to avoid repetition?

**Comment by richardcase** [comment]
@Danil-Grigorev - is there anything outstanding on your side for this?

**Comment by furkatgofurov7** [comment]
@Danil-Grigorev this LGTM, can you drop second commit https://github.com/rancher-sandbox/rancher-turtles/pull/114/commits/421c3320a0324c95bad33ab2be77afb46168cda5 as it looks redundant?

**Comment by Danil-Grigorev** [comment]
Should be all good now @richardcase @furkatgofurov7

File: docs/adr/0009-use-structured-proxy-types.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - [Consequences](#consequences)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# 9. Use structured proxy types
+
+- Status: proposed
+- Date: 2023-08-09
+- Authors: @Danil-Grigorev
+- Deciders: @richardcase @alexander-demicev @salasberryfin @Danil-Grigorev @mjura
```
Comment (furkatgofurov7) [inline docs/adr/0009-use-structured-proxy-types.md:16]: not in the list, but that is fine 😄

**Review by furkatgofurov7 — APPROVED** [review]

**Review by furkatgofurov7 — APPROVED** [review]

**Review by alexander-demicev — APPROVED** [review]

**Comment by alexander-demicev** [comment]
@Danil-Grigorev can we ignore the CI failure for this PR?

**Pull request merged**